pkg/devicetwin/dtmanager/membership_test.go

```diff
@@ -0,0 +1,352 @@
+		Content: "invalidmsg",
+	}
+
+	value, err := dealMembershipDetail(dtc, "t", m)
+	assert.Error(t, err)
+	assert.Equal(t, errors.New("assertion failed"), err)
+	assert.Equal(t, nil, value)
+}
+
+func TestDealMembershipDetailInvalidContent(t *testing.T) {
+
+	dtc := &dtcontext.DTContext{
+		DeviceList: &sync.Map{},
+		GroupID:    "1",
+	}
+	var cnt []uint8
+	cnt = append(cnt, 1)
+	var m = &model.Message{
+		Content: cnt,
+	}
+
+	value, err := dealMembershipDetail(dtc, "t", m)
+	assert.Error(t, err)
+	assert.Equal(t, nil, value)
+}
+
+func TestDealMembershipDetailValid(t *testing.T) {
+	dtc := &dtcontext.DTContext{
+		DeviceList: &sync.Map{},
+		Mutex:      &sync.Mutex{},
+		GroupID:    "1",
+	}
+
+	payload := dttype.MembershipUpdate{AddDevices: []dttype.Device{{ID: "DeviceA", Name: "Router",
+		State: "unknown"}}, BaseMessage: dttype.BaseMessage{EventID: "eventid"}}
+	content, _ := json.Marshal(payload)
+	var m = &model.Message{
+		Content: content,
+	}
+	_, err := dealMembershipDetail(dtc, "t", m)
+	assert.NoError(t, err)
+}
+
+func TestDealMembershipUpdatedEmptyMessage(t *testing.T) {
+	dtc := &dtcontext.DTContext{
+		DeviceList: &sync.Map{},
+		GroupID:    "1",
+	}
+	value, err := dealMembershipDetail(dtc, "t", "invalid")
+	assert.Error(t, err)
+	assert.Equal(t, errors.New("msg not Message type"), err)
+	assert.Equal(t, nil, value)
+}
+
+func TestDealMembershipUpdatedInvalidMsg(t *testing.T) {
+
+	dtc := &dtcontext.DTContext{
+		DeviceList: &sync.Map{},
+		GroupID:    "1",
+	}
+
+	var m = &model.Message{
+		Content: "invalidmessage",
+	}
+
+	value, err := dealMembershipUpdated(dtc, "t", m)
+	assert.Error(t, err)
+	assert.Equal(t, errors.New("assertion failed"), err)
+	assert.Equal(t, nil, value)
+}
+func TestDealMembershipUpdatedInvalidContent(t *testing.T) {
+
+	dtc := &dtcontext.DTContext{
+		DeviceList: &sync.Map{},
+		GroupID:    "1",
+	}
+
+	var cnt []uint8
+	cnt = append(cnt, 1)
+	var m = &model.Message{
+		Content: cnt,
+	}
+
+	value, err := dealMembershipUpdated(dtc, "t", m)
+	assert.Error(t, err)
+	assert.Equal(t, nil, value)
+}
+
+func TestDealMembershipUpdatedValidAddedDevice(t *testing.T) {
+	initMocks(t)
+	ormerMock.EXPECT().Begin().Return(nil).Times(6)
+	ormerMock.EXPECT().Insert(gomock.Any()).Return(int64(1), nil).Times(3)
+	ormerMock.EXPECT().Commit().Return(nil).Times(1)
+	querySeterMock.EXPECT().Filter(gomock.Any(), gomock.Any()).Return(querySeterMock).Times(1)
+	dtc := &dtcontext.DTContext{
+		DeviceList:  &sync.Map{},
+		DeviceMutex: &sync.Map{},
+		Mutex:       &sync.Mutex{},
+		GroupID:     "1",
+	}
+
+	payload := dttype.MembershipUpdate{AddDevices: []dttype.Device{{ID: "DeviceA", Name: "Router",
+		State: "unknown"}}, BaseMessage: dttype.BaseMessage{EventID: "eventid"}}
+	content, _ := json.Marshal(payload)
+	var m = &model.Message{
+		Content: content,
+	}
+	_, err := dealMembershipUpdated(dtc, "t", m)
+	assert.NoError(t, err)
+}
+
+func TestDealMembershipUpdatedValidRemovedDevice(t *testing.T) {
+	initMocks(t)
```

Comment:
I guess init mocks need to be called only once, once initialised mocks can be used in all functions.

Reply:
Removed from other function and initialised in one place.